report_py3o/py3o_parser.py

```diff
@@ -0,0 +1,217 @@
+def defautl_extend(report_xml, localcontext):
+    # add the base64decode function to be able do decode binary fields into
+    # the template
+    localcontext['b64decode'] = b64decode
+
+
+class Py3oParser(report_sxw):
+    """Custom class that use Py3o to render libroffice reports.
+        Code partially taken from CampToCamp's webkit_report."""
+
+    def get_template(self, report_obj):
+        """private helper to fetch the template data either from the database
+        or from the default template file provided by the implementer.
+
+        ATM this method takes a report definition recordset
+        to try and fetch the report template from database. If not found it
+        will fallback to the template file referenced in the report definition.
+
+        @param report_obj: a recordset representing the report defintion
+        @type report_obj: openerp.model.recordset instance
+
+        @returns: string or buffer containing the template data
+
+        @raises: TemplateNotFound which is a subclass of
+        openerp.exceptions.DeferredException
+        """
+
+        tmpl_data = None
+
+        if report_obj.py3o_template_id and report_obj.py3o_template_id.id:
+            # if a user gave a report template
+            tmpl_data = b64decode(
+                report_obj.py3o_template_id.py3o_template_data
+            )
+
+        elif report_obj.py3o_template_fallback:
+            tmpl_name = report_obj.py3o_template_fallback
+            flbk_filename = None
+            if report_obj.module:
+                # if the default is defined
+                flbk_filename = pkg_resources.resource_filename(
+                    "openerp.addons.%s" % report_obj.module,
+                    tmpl_name,
+                )
+            elif os.path.isabs(tmpl_name):
+                # It is an absolute path
+                flbk_filename = os.path.normcase(os.path.normpath(tmpl_name))
+            if flbk_filename and os.path.exists(flbk_filename):
+                # and it exists on the fileystem
+                with open(flbk_filename, 'r') as tmpl:
+                    tmpl_data = tmpl.read()
+
+        if tmpl_data is None:
+            # if for any reason the template is not found
+            raise TemplateNotFound(
+                _(u'No template found. Aborting.'),
+                sys.exc_info(),
+            )
+
+        return tmpl_data
+
+    def _extend_parser_context(self, parser_instance, report_xml):
+        # add default extenders
+        for fct in _extender_functions.get(None, []):
+            fct(report_xml, parser_instance.localcontext)
+        # add extenders for registered on the template
+        xml_id = report_xml.get_external_id().get(report_xml.id)
+        if xml_id in _extender_functions:
+            for fct in _extender_functions[xml_id]:
+                fct(report_xml, parser_instance.localcontext)
+
+    def create_single_pdf(self, cr, uid, ids, data, report_xml, context=None):
```

Comment:
@lasley here we are in the parser not in an odoo model. Therefore the new api is applicable and moreover our class extends a pure python class and we override only some methods... so we must keep the same method signature